src/Microsoft.AspNet.PlatformHandler/HttpPlatformHandlerMiddleware.cs

```diff
@@ -0,0 +1,85 @@
+
+using System;
+using System.Globalization;
+using System.Net;
+using System.Security.Principal;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+using Microsoft.Framework.Internal;
+using Microsoft.Framework.Primitives;
+
+namespace Microsoft.AspNet.PlatformHandler
+{
+    public class HttpPlatformHandlerMiddleware
+    {
+        private const string XForwardedForHeaderName = "X-Forwarded-For";
+        private const string XForwardedProtoHeaderName = "X-Forwarded-Proto";
+        private const string XIISWindowsAuthToken = "X-IIS-WindowsAuthToken";
+        private const string XOriginalProtoName = "X-Original-Proto";
+        private const string XOriginalIPName = "X-Original-IP";
+
+        private readonly RequestDelegate _next;
+
+        public HttpPlatformHandlerMiddleware(RequestDelegate next)
+        {
+            _next = next;
+        }
+
+        public Task Invoke(HttpContext httpContext)
+        {
+            var xForwardProtoHeaderValue = httpContext.Request.Headers[XForwardedProtoHeaderName];
+            if (!string.IsNullOrEmpty(xForwardProtoHeaderValue))
+            {
+                if (!string.IsNullOrEmpty(httpContext.Request.Scheme))
+                {
+                    httpContext.Request.Headers[XOriginalProtoName] = httpContext.Request.Scheme;
+                }
+                httpContext.Request.Scheme = xForwardProtoHeaderValue;
+            }
+            
+            var xForwardedForHeaderValue = httpContext.Request.Headers.GetCommaSeparatedValues(XForwardedForHeaderName);
+            if (xForwardedForHeaderValue != null && xForwardedForHeaderValue.Length > 0)
+            {
+                IPAddress ipFromHeader;
+                if (IPAddress.TryParse(xForwardedForHeaderValue[0], out ipFromHeader))
+                {
+                    var remoteIPString = httpContext.Connection.RemoteIpAddress?.ToString();
+                    if (!string.IsNullOrEmpty(remoteIPString))
+                    {
+                        httpContext.Request.Headers[XOriginalIPName] = remoteIPString;
+                    }
+                    httpContext.Connection.RemoteIpAddress = ipFromHeader;
+                }
+            }
+
+            var xIISWindowsAuthToken = httpContext.Request.Headers[XIISWindowsAuthToken];
+            int hexHandle;
+            if (!StringValues.IsNullOrEmpty(xIISWindowsAuthToken)
+                && int.TryParse(xIISWindowsAuthToken, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out hexHandle))
+            {
+                var handle = new IntPtr(hexHandle);
+                var winIdentity = new WindowsIdentity(handle);
+                // WindowsIdentity just duplicated the handle so we need to close the original.
+                NativeMethods.CloseHandle(handle);
+
+                httpContext.Response.RegisterForDispose(winIdentity);
```

Comment:
Why do this instead of handle it on the way out (via try finally)? It means that middleware registered before this will see the WindowsIdentity on the way out. Isn't that weird?